src/context/authStore/bucketStorageHelper.ts

```diff
@@ -0,0 +1,120 @@
+import * as SecureStore from "expo-secure-store";
+
+/** size limit of each entry in SecureStore in bytes */
+const SIZE_LIMIT = 2048;
+
+/**
+ * Saves a string to SecureStore, respecting the
+ * [2048 size limit](https://docs.expo.io/versions/v42.0.0/sdk/securestore/)
+ * by splitting the serialised string up and saving it under multiple keys.
+ *
+ * The string to be saved is split into buckets (substring), with each bucket
+ * not exceeding the size limit. Each bucket is then saved under it's own key
+ * in SecureStore, starting with "{@link storageKey}_0" and then overflowing
+ * into "{@link storageKey}_1", "*_2", etc.
+ *
+ * Each bucket is compared with the original value at the bucket (based on
+ * {@link oldValue}) and only buckets that have changed are updated, the rest
+ * are left untouched.
+ * If {@link newValue} takes up less buckets than {@link oldValue}, extra
+ * buckets are deleted from SecureStore.
+ * The exception for the first bucket is that an empty string is counted as
+ * data and will be stored in the first bucket instead of deleting it,
+ * differentiating an empty string from no data.
+ *
+ * Note that the function does not read the existing value stored and takes
+ * reference from {@link oldValue} instead.
+ *
+ * All storage calls are done asynchronously without blocking the caller.
+ *
+ * @param newValue new value to be stored
+ * @param oldValue previous value stored
+ * @param bucketSize number of keys per bucket
+ * @param storageKey key to be used for storing
+ */
+export const saveToStoreInBuckets: (
+  storageKey: string,
+  newValue: string,
+  oldValue: string | null
+) => void = (storageKey, newValue, oldValue) => {
+  oldValue = oldValue ?? "";
+
+  // iterate the longer length in case we have to clear buckets
+  const compareLength = Math.max(newValue.length, oldValue.length);
+  for (let bucketNo = 0; bucketNo * SIZE_LIMIT < compareLength; bucketNo++) {
+    const newValueBucketString = newValue.slice(
+      bucketNo * SIZE_LIMIT,
+      (bucketNo + 1) * SIZE_LIMIT
+    );
+    const oldValueBucketString = oldValue.slice(
+      bucketNo * SIZE_LIMIT,
+      (bucketNo + 1) * SIZE_LIMIT
+    );
+
+    if (newValueBucketString !== oldValueBucketString) {
+      if (newValueBucketString === "" && bucketNo !== 0) {
+        // if the new bucket is empty, delete the key
+        // however, if this is the first bucket, user is saving an empty string, so do not delete
+        SecureStore.deleteItemAsync(storageKey + "_" + bucketNo);
```

Comment:
I am not too sure which is better and whether it will solve the following hypothetical problem I thought of. Although from a reusable code standpoint, it does make sense to `await` for it and leave it up to the caller if they want to block or carry on.

> Possible way to break the saving logic:
> 
> Currently, we are required to save multiple keys at once. There are two ways to do this, one is to lock the saving logic (the Effect in authStore.tsx:65) and the other is to just run everything without locking.
> 
> With locking: no other updates can happen while we are saving the current value. If there are any changes during this time, we will lose them unless we set up some kind of queue for changes.
> Without locking: if another update happens while we are saving the current value, the async calls from both might run concurrently and override each other.
> 
> Caveat: this requires the two updates to happen really close together, which is unlikely as it means a user needs to add a new campaign/logout successively really quickly.

Reply:
If we look at `saveToStoreInBuckets` independently, a caller will expect it to be synchronous, i.e. once that fn returns, the store would have been updated. I suggest we use `Promise.all` to ensure that all the async calls are resolved before the fn is returned. Note that the use of `Promise.all` immediately rejects if one of the promises fails.

For the useEffect in authStore, what you pointed out about a race condition with `saveToStoreInBuckets` is indeed possible, though unlikely. The locking can look something like this: https://github.com/facebook/react/issues/14326#issuecomment-472043812 However, this complexity isn't worthwhile here as it's unlikely to happen.